### PR TITLE
blockVolume: fix over allocation of small volumes

### PR DESCRIPTION
**Pull request by aesteve-rh**
For block sparse volumes, if no initial_size is
defined, Vdsm allocates one chunk (2.5G), without
considering the virtual size of the volume.
This results in wasted space if the requested
volume is smaller.

Instead, we can allocate the capacity when
it is smaller than one chunk, to avoid wasting
space for small volumes.
Note than any size smaller than one extent
(128M) will be rounded to the extent size.

Bug-Url: https://bugzilla.redhat.com/2094576
Signed-off-by: Albert Esteve <aesteve@redhat.com>

### CODE REVIEW

File: lib/vdsm/storage/blockVolume.py
Comment (nirs) [inline lib/vdsm/storage/blockVolume.py:371]: This will allocate 128M if you try to create 512M volume. What we really want is to allocate one chunk, but limit the chunk size to the capacity:

    alloc_size = max(chunk_size_mb * MiB, capacity)

The comment is also not helpful, it describes the code instead of why the code exist. Since this is trivial change, I don't think a comment is needed.

We need to add new test verifying that we allocated the right size in this case. The best way to do such changes it to add first the test, reproducing the bug. The test must fail to prove that the test is correct. Then add the fix which will fix the failure, proving that the fix is correct.

If the fix causes other tests to fail, it is best to fix the other failing tests assuming minimal allocation of one chunk (even if capacity < chunk) before you add the fix.

File: tests/storage/merge_test.py
Comment (nirs) [inline tests/storage/merge_test.py:371]: Can you explain this change?

File: tests/storage/merge_test.py
Comment (nirs) [inline tests/storage/merge_test.py:508]: Why did you add size here?

File: tests/storage/merge_test.py
Comment (nirs) [inline tests/storage/merge_test.py:525]: The assumption that we don't need to reduce after merge is based on wrong assumption
about creating volume with one chunk. If you create the volume with initial_size=capacity the volume will be reduced after merge.

Changing merge logic because because allocation bug was fixed is not good. We want to keep the existing behavior and fix the allocation bug.

File: tests/storage/merge_test.py
Comment (nirs) [inline tests/storage/merge_test.py:566]: Not clear why you added size here.

**Review by nirs — CHANGES_REQUESTED** [review]

File: lib/vdsm/storage/blockVolume.py
Comment (aesteve-rh) [reply lib/vdsm/storage/blockVolume.py:371]: `What we really want is to allocate one chunk, but limit the chunk size to the capacity` -> that would be min, no?
```python
    alloc_size = min(chunk_size_mb * MiB, capacity)
```
And that would make the allocation to be smaller than an extent if the capacity is smaller (as is the case in BZ). I can max it later with the align size, maybe?

And in this case, we already had a test covering the usecase in `blockvolume_test`: `TestBlockVolumeSize`.

File: tests/storage/merge_test.py
Comment (aesteve-rh) [reply tests/storage/merge_test.py:508]: With this implementation, capacities smaller than 2.5G (chunk size), allocated only one extent, and weren't reducing after merge. I changed it so that the size was bigger than the chunk size to get the reduce call in the test.

File: lib/vdsm/storage/blockVolume.py
Comment (nirs) [reply lib/vdsm/storage/blockVolume.py:371]: Yes I sometimes get min/max wrong, and this should be min().

Getting alloc_size < align_size is fine! lvm rounds the value up to the extent size
and create volume code adjust the size properly before writing the metadata.

You can add max(align_size, alloc_size) to align the value before calling lvm if you
think it makes the code more clear, but this may be needed in other cases, so best
do this later in all cases and not only for the case fixed here.

File: tests/storage/merge_test.py
Comment (nirs) [reply tests/storage/merge_test.py:508]: This changes are best done in a separate patch, with commit message explaining why the change is needed. This makes the actual change smaller and safer and keeps the rule that every commit should have exactly one logical change.

File: lib/vdsm/storage/blockVolume.py
Comment (aesteve-rh) [reply lib/vdsm/storage/blockVolume.py:371]:
```
Getting alloc_size < align_size is fine! lvm rounds the value up to the extent size
and create volume code adjust the size properly before writing the metadata.
```
That's good to know! Then I think the max is not really needed. Maybe a comment or a specific test would suffice. Thanks for the clarification.

File: tests/storage/merge_test.py
Comment (aesteve-rh) [reply tests/storage/merge_test.py:525]: `If you create the volume with initial_size=capacity the volume will be reduced after merge.` Unfortunately, that's not what I observe in the test.

The test creates block volumes using `calculate_volume_alloc_size`, which results in initial_size=capacity.
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L333

On finalize, we obtain the optimal_size and the actual_size (in this case, actual_size=capacity):

Optimal_size:
In this test we create COW volumes, so we obtain the optimal size in `optimal_cow_size`:
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L421
In this case what prevails for the optimal size calculation is this line:
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L448
Which results in `optimal_size = int(capacity * 1.1)` <- this is not in the code, but is closely equivalent to lines 448 and 449 (it also aligns the result to the extent size).

Therefore, optimal_size is always bigger than actual_size (because `int(capacity * 1.1) > capacity`), and the volume is never reduced:
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/merge.py#L286
Few actual examples
- 2 GiB volumes: 2.2 GiB optimal_size > 2 GiB actual_size.
- 1 GiB volumes: 1.1 GiB optimal_size > 1 GiB actual_size.
- 1 MiB volumes: 128 MiB optimal_size = 128 MiB actual_size

Comment (aesteve-rh) [reply tests/storage/merge_test.py:525]: On the other hand, for sizes bigger than a chunk:

`calculate_volume_alloc_size` will give an initial_size=chunk_size, for any chosen capacity in the test.

In `optimal_cow_size` the most relevant part is this one:
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L445
Which comes from `required_size (current data offset) + chunk_size` (aligned to the next extent_size).

Thus, for all these cases, we always obtain the same result:
- 3 GiB volumes (or any volume > 2.5 GiB): 2.625 GiB optimal_size > 2.5 GiB chunk size

And volume is never reduced.

File: tests/storage/merge_test.py
Comment (aesteve-rh) [reply tests/storage/merge_test.py]: I have removed this change, was not required after changing the `alloc_size` to the capacity in:
```python
    alloc_size = min(chunk_size_mb * MiB, capacity)
```

File: tests/storage/merge_test.py
Comment (aesteve-rh) [reply tests/storage/merge_test.py]: Finally removed this change.

File: tests/storage/blockvolume_test.py
Comment (nirs) [inline tests/storage/blockvolume_test.py]: Not sure this is better - it is shorter code but the current code is about the actual vdsm apis used, so it is more clear what is going on. Adding a constant hide things.

File: tests/storage/blockvolume_test.py
Comment (nirs) [inline tests/storage/blockvolume_test.py:59]: Using fake_env is very likey to create wrong tests. This env try to fake a block device with file and fake lvm, and may give wrong results.

This should be really tested in blocksd_test, using real storage domain and running real createVolume() calls.

File: tests/storage/merge_test.py
Comment (nirs) [inline tests/storage/merge_test.py:525]: Marking test as xfail because it does not work after the fix does not make sense.

If the tests does not work now change it to test the real behavior. I wonder why this test succeeded before. We should test here real volume size (e.g. 10g).

The fix should not affect merge tests, it it breaks them the tests probably not testing real use case.

File: tests/storage/merge_test.py
Comment (nirs) [reply tests/storage/merge_test.py:525]: Testing 2G volume is not interesting, real disk are much bigger - changing the test to test 10g volume.

It is possible that this test used small volume size, and relied on the wrong allocation (2.5g) to test the reduce, and now when we create 2g (or 2.2) we don't need to reduce.

Comment (aesteve-rh) [reply tests/storage/merge_test.py:525]: As I tried to expose in this comment: https://github.com/oVirt/vdsm/pull/293#discussion_r941206435

Any volume size bigger than a chunk will not get reduced, since:
- vdsm allocates one chunk -> `actual_size = chunk_size`
- optimal_size will be `chunk_size + offset` for leaf volumes
- optimal_size > actual_size

It definitively relied in the over-allocation to trigger the reduce. The test only creates the volumes and invokes `finalize()`, so if `calculate_volume_alloc_size` does not allocate more than the `optimal_size` it just never gets reduced.

File: tests/storage/blockvolume_test.py
Comment (aesteve-rh) [reply tests/storage/blockvolume_test.py]: For a single occurrence is ok, for multiple occurrences it made the permutations more difficult to read.

I think the constant declared in consecutive lines with the `CONFIG`, makes it clear that the permutations use the direct size, and the test shall obtain the same size from the `CONFIG` structure that is monkeypatched.

Which btw made me realize that there is a typo in the `CONFIG`, `volume_utilizzation_precent` is wrong.

File: tests/storage/blockvolume_test.py
Comment (aesteve-rh) [reply tests/storage/blockvolume_test.py:59]: I'm ok with adding a `blocksd_test`, but I'd still keep this one, it is consistent with the other tests in the module, checking size, allocation, max_size, and optimal_size with the same strategy and code style.

File: tests/storage/merge_test.py
Comment (aesteve-rh) [reply tests/storage/merge_test.py:525]: That's why I removed it in the first place, xfail just shows that it is indeed failing in the pipelines.

It was working because we were overallocating for small volumes and the test levarages that. Bigger volumes (e.g., 10g) will not work, with current code or previous code.

In general what is making this test fail (vs other similar test) is that leaf volumes have an increased optimal_size, which makes it consistently bigger than the allocated size: 
https://github.com/aesteve-rh/vdsm/blob/aesteve/small-volumes-blocksd/lib/vdsm/storage/blockVolume.py#L432

File: tests/storage/merge_test.py
Comment (aesteve-rh) [reply tests/storage/merge_test.py:525]: One way to make it pass is forcing over-allocation through fakery, e.g.:
```python
        monkeypatch.setattr(
            blockVolume.BlockVolumeManifest,
            "calculate_volume_alloc_size",
            lambda *x: 4 * GiB)
```
I guess we could also write and then free some space in the volume, so that it can be reduced.

But probably this test does not make sense anymore and can be removed.
- Sparse leaf volumes will never be over-allocated with the bug-fix in this branch, so will never reduce.
- Preallocated volumes are not allowed to be reduced.

File: tests/storage/merge_test.py
Comment (aesteve-rh) [reply tests/storage/merge_test.py:525]: I'll close this thread to not to diverge from the other [thread](https://github.com/oVirt/vdsm/pull/293#discussion_r938228290), which is a related discussion (aka what to do with the test that now fails no matter which size we use).

File: tests/storage/merge_test.py
Comment (aesteve-rh) [reply tests/storage/merge_test.py]: Fixed with https://github.com/oVirt/vdsm/pull/316

File: tests/storage/blockvolume_test.py
Comment (barpavel) [inline tests/storage/blockvolume_test.py:107]: Not sure whether the comment should be updated (in lines 105 & 107): `config.volume_utilization_chunk_mb` --> `CHUNK_SIZE_MB`.

File: tests/storage/blocksd_test.py
Comment (barpavel) [inline tests/storage/blocksd_test.py:883]: Suggestion:
`vol_id` --> `vol_uuid`.

File: tests/storage/blockvolume_test.py
Comment (barpavel) [inline tests/storage/blockvolume_test.py:57]: Maybe better use `img_uuid` & `vol_uuid`?

**Review by barpavel — APPROVED** [review]

File: tests/storage/blockvolume_test.py
Comment (aesteve-rh) [reply tests/storage/blockvolume_test.py:57]: ok, this is actually moved code, not created new, so I moved it as is. But it is ok, I can update the name of these variables in the process.

File: tests/storage/blockvolume_test.py
Comment (aesteve-rh) [reply tests/storage/blockvolume_test.py:58]: Same as before, I moved this line as is. I'm fine with a small update though :)

File: tests/storage/blockvolume_test.py
Comment (aesteve-rh) [reply tests/storage/blockvolume_test.py:107]: Both are true, and leaving the config in the comment hints for the source of the `CHUNK_SIZE_MB`, so I think we can keep it as is?

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:58]: Any reason you use blockSD.config instead of importing config and fetching the value from the config?

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py]: vdsm over-allocates small volumes

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:877]: test_cow_small_volume?

A comment explaining what is small volume can help (e.g. capacity < chunk size)

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:875]: Not sure why we need so many tests. The interesting cases are:
- capacity == chunk size -> allocate capacity
- capacity == chunk size - 1 -> allocate capacity
- capacity < extent size -> allocate extent size
- capacity > chunk size -> allocated chunk (but this must be already covered by other tests)

(Allocations are always rounded up to extent size)

These tests are slow, so minimizing the number of tests is very important. Using parameters for this tests is really bad - we will create the loop devices and storage domain for creating a single volume for every test!

All the cases should be tested with the same storage domain. You can create one test that create 3 volumes instead of using parameters.

File: tests/storage/blockvolume_test.py
Comment (nirs) [inline tests/storage/blockvolume_test.py:88]: This repeats the blocksd tests in a bad way. I would delete the duplicate tests here instead of wasting time on the fake storage.

File: tests/storage/blocksd_test.py
Comment (aesteve-rh) [reply tests/storage/blocksd_test.py:58]: No reason other than to avoid having the extra import. But does no harm and will make it more clear, so I will change it. As it is, the real path to config gets obscured.

File: tests/storage/blocksd_test.py
Comment (aesteve-rh) [reply tests/storage/blocksd_test.py]: Ok, I did put the three usecases together, avoided using parametrized test, and commented the whole test to help to read it.

File: tests/storage/blockvolume_test.py
Comment (aesteve-rh) [reply tests/storage/blockvolume_test.py]: Removed.

File: tests/storage/blockvolume_test.py
Comment (aesteve-rh) [reply tests/storage/blockvolume_test.py:77]: I removed the constant because you seemed not convinced before, and after removing all the additional cases and test, my justification for having it (same as yours, more clear test data) felt not as relevant anymore.

I can bring it back, no problem. Lets will wait for the ost test to finish, and I will push the update.

**Review by nirs — APPROVED** [review]

**Comment by aesteve-rh** [comment]
Last OST run finished successfully: https://rhv-devops-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/ds-ost-baremetal-gh/1619/

@nirs or @mz-pdm this can be merged if everything else looks ok. And then we'll be ready for the next build :)

**Pull request merged**
